### ISSUE TITLE
Tracking API: track Emoji correctly in page URLs and others

### ISSUE THREAD

**Issue by mattab** [matomo-org/matomo#8790]
Goal: correctly track emoji and all utf8 4-byte characters. This requires a change from our table charset to `utf8mb4`. See discussion in #7766

**Comment by shivateja-madipalli** [comment]
Not sure, whether this is the right place to ask a question.
I am new to Piwik and got stuck at this point and want some help for the community. 
Piwik provides API to get the current visitor id from the browser, 

Piwik.getAsyncTracker().getVisitorInfo(); 
_paq.push([ function() { visitor_id = this.getVisitorId(); }]); 

but these api's are giving different visitor id on every refresh (i.e., on every page visit) none of the ids provided is not equal to visitor profile's id in piwik dashboard. 

If there for sure a way to get visitor id from Piwik API itself irrespective of clearing the browse cookies or not, then below question is not really necessary. 

Thus, to fix this i have been retrieving the visitor id value from Piwik cookie "_pk_id.2.89be" (first value is the visitor id) stored in client browser. 
But Now, the actual problem is, When ever client browser cookies are cleared then visitor id in the cookie is changing, How to get primarily created visitor id?? 

Inputs: Piwik is still able to track on the basis of this new visitor id and appending the tracking data under the same visitor id, so, there is some where in piwik or client browser that for every first visit from a finger print (IP + System echo system) to the website, piwik is assigning a visitor id, and every other visit's tracking is being appended to the same visitor ID.

**Comment by huan086** [comment]
Do use utf8mb4_unicode_520_ci for the charset for correct sorting. E.g. https://bugs.mysql.com/bug.php?id=2906

**Comment by huan086** [comment]
Quick and dirty hack to enable emoji

In piwik\config\config.ini.php, after
```ini
adapter = "MYSQLI"
```
add
```ini
charset = "utf8mb4"
```

Run the following SQL
```sql
ALTER SCHEMA `piwik` DEFAULT CHARACTER SET utf8mb4 DEFAULT COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_access` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_archive_blob_yyyy_mm` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci; -- repeat this for all the tables, with yyyy and mm replaced
ALTER TABLE `piwik`.`piwik_archive_numeric_yyyy_mm` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci; -- repeat this for all the tables, with yyyy and mm replaced
ALTER TABLE `piwik`.`piwik_goal` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_log_action` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_log_conversion` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_log_conversion_item` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_log_link_visit_action` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_log_profiling` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_log_visit` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_logger_message` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_option` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_plugin_setting` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_report` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_segment` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_sequence` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_session` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_site` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_site_setting` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_site_url` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_user` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_user_dashboard` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_user_language` CHARACTER SET=utf8mb4, COLLATE=utf8mb4_unicode_520_ci;

ALTER TABLE `piwik`.`piwik_access` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_archive_blob_yyyy_mm` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci; -- repeat this for all the tables, with yyyy and mm replaced
ALTER TABLE `piwik`.`piwik_archive_numeric_yyyy_mm` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci; -- repeat this for all the tables, with yyyy and mm replaced
ALTER TABLE `piwik`.`piwik_goal` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_log_action` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_log_conversion` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_log_conversion_item` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_log_link_visit_action` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_log_profiling` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_log_visit` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_logger_message` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_option` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_plugin_setting` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_report` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_segment` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_sequence` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_session` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_site` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_site_setting` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_site_url` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_user` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_user_dashboard` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;
ALTER TABLE `piwik`.`piwik_user_language` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci;

**Comment by saqib16** [comment]
Hi,

We have updated to Piwik 3.0.2 and PHP 7.0.16 and getting following error:

Error in Piwik (tracker): Error query: SQLSTATE[HY000]: General error: 1366 Incorrect string value: '\xD0_\xD0\xBB\xD0\xB5...' for column 'name' at row 1 In query: INSERT INTO piwik_log_action (name, hash, type, url_prefix) VALUES (?,CRC32(?),?,?) Parameters: array ( 0 => '/products/Ð¡Ñ‡ÐµÑ‚Ñ‡Ð¸ÐºÐ¸ Ð_Ð»ÐµÐºÑ‚Ñ€Ð¾Ñ_Ð½ÐµÑ€Ð³Ð¸Ð¸/?cid=91701', 1 => '/products/Ð¡Ñ‡ÐµÑ‚Ñ‡Ð¸ÐºÐ¸ Ð_Ð»ÐµÐºÑ‚Ñ€Ð¾Ñ_Ð½ÐµÑ€Ð³Ð¸Ð¸/?cid=91701', 2 => 1, 3 => 0, )

Any guess/solution appreciated?

**Comment by sgiehl** [comment]
this should be fixed when #9785 is done
